### PR TITLE
Arithmetic Pattern

### PR DESCRIPTION
**Pull request by Darredevil**
Patch for #5397 
Work in progress ,  please ignore warnings and comments for now

### CODE REVIEW

File: libr/core/cmd_search_rop.c [hunk 1]
```diff
@@ -52,9 +52,101 @@ static RList* get_constants (const char *str) {
...
+	} else if (!strcmp (op, "&=")) {
+		return src1 & src2;
+	} else if (!strcmp (op, "^=")) {
+		return src1 ^ src2;
+	} else if (!strcmp (op, "|=")) {
+		return src1 | src2;
+	}
+
+	// fail value
+	return -1337;
```
Comment (crowell) [inline libr/core/cmd_search_rop.c]: Have this take a ptr to a value, and return a bool for success instead.

File: libr/core/cmd_search_rop.c [hunk 2]
```diff
@@ -52,9 +52,101 @@ static RList* get_constants (const char *str) {
 	return list;
 }
 
+static bool isFlag (const char *f) {
+	const char *flags = "zf,pf,sf,cf,of";
+
+	if (strstr (flags, f)) {
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: wat this is not portable and wrong way to check if its a flag. you can take this info from rreg

File: libr/core/cmd_search_rop.c [hunk 3]
```diff
@@ -52,9 +52,101 @@ static RList* get_constants (const char *str) {
...
+		return true;
+	} else {
+		return false;
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: no need to else if the first block returns already

File: libr/core/cmd_search_rop.c [hunk 4]
```diff
@@ -52,9 +52,101 @@ static RList* get_constants (const char *str) {
...
+	}
+}
+
+// TODO: hideous code, must rafactor
+// binary op
+static ut64 simulate_op (const char *op, ut64 src1, ut64 src2, ut64 old_src1, ut64 old_src2) {
+
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: empty line

File: libr/core/cmd_search_rop.c (hunk 1 above)
Comment (radare) [inline libr/core/cmd_search_rop.c]: dont mix error and data channels. return the value as a ref and return bool for example

Comment (radare) [inline libr/core/cmd_search_rop.c]: if you use magic constants name them in a const or define

File: libr/core/cmd_search_rop.c [hunk 5]
```diff
@@ -52,9 +52,101 @@ static RList* get_constants (const char *str) {
...
+	}
+
+	iter_reg = regs->head;
+	reg_item = iter_reg->data;
+	r_reg_arena_pop (core->dbg->reg);
+	r_list_foreach (core->io->sections, iter_io, sec) {
+		if (sec->rwx & 1) {
+			addr = sec->vaddr;
+			// r_cons_printf ("RWX addr = 0x%"PFMT64x"\n", addr);
+			if (iter_reg && reg_item) {
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: cant we just move this check outside the loop, right after setting those vars? and drop the return statement here

File: libr/core/cmd_search_rop.c [hunk 6]
```diff
@@ -235,6 +338,137 @@ static char* rop_classify_mov (RCore *core, RList *ropList) {
...
+		// }
+
+		// init regs with random values
+		// fillRegisterValues (core);
+		head = r_reg_get_list (core->dbg->reg, 0);
+		if (!head) {
+			return NULL;
+		}
+		r_list_foreach (head, iter_dst, item_dst) {
+			int r = r_num_rand (30000);
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: randomness is scary

File: libr/core/cmd_search_rop.c [hunk 7]
```diff
@@ -235,6 +338,137 @@ static char* rop_classify_mov (RCore *core, RList *ropList) {
...
+				}
+
+				r_list_foreach (head, iter_src2, item_src2) {
+					ut64 value_src2, diff_src2;
+
+					value_src2 = r_reg_get_value (core->dbg->reg, item_src2);
+					r_reg_arena_swap (core->dbg->reg, false);
+					diff_src2 = r_reg_get_value (core->dbg->reg, item_src2);
+					r_reg_arena_swap (core->dbg->reg, false);
+
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: empty line

**Comment by radare** [comment]
Please fix all the comments

File: libr/core/cmd_search_rop.c (hunk 7 above)
Comment (Darredevil) [inline libr/core/cmd_search_rop.c]: fixed, false postive now

File: libr/core/cmd_search_rop.c (hunk 6 above)
Comment (Darredevil) [inline libr/core/cmd_search_rop.c]: pending next patch and args parsing

File: libr/core/cmd_search_rop.c [hunk 8]
```diff
@@ -52,9 +52,139 @@ static RList* get_constants (const char *str) {
 	return list;
 }
 
+static bool isFlag (RRegItem *reg) {
+	const char *type = r_reg_get_type (reg->type);
+
+	if (!strcmp(type, "flg"))
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: what about return !strcmp (type, "flg")); // remember the spaces before (

File: libr/core/cmd_search_rop.c [hunk 9]
```diff
@@ -52,9 +52,139 @@ static RList* get_constants (const char *str) {
...
+	}
+	if (!strcmp (op, "^=")) {
+		*result = src1 ^ src2;
+		return true;
+	}
+	if (!strcmp (op, "|=")) {
+		*result = src1 | src2;
+		return true;
+	}
+
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: empty line

File: libr/core/cmd_search_rop.c [hunk 10]
```diff
@@ -69,9 +199,9 @@ static char* rop_classify_constant (RCore *core, RList *ropList) {
 	}
 
 	r_list_foreach (ropList, iter_r, esil_str) {
-		if (strchr (esil_str, '[')) { // avoid MEM read/write for now
-			return NULL;
-		}
+		// if (strchr (esil_str, '[')) { // avoid MEM read/write for now
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: remove commented code

File: libr/core/cmd_search_rop.c [hunk 11]
```diff
@@ -200,12 +331,18 @@ static char* rop_classify_mov (RCore *core, RList *ropList) {
 				continue;
 			}
 
+			// you never mov into flags
+			// if (isFlag (item_dst->name, core->dbg->reg)) {
```
Comment (radare) [inline libr/core/cmd_search_rop.c]: remove that line

File: libr/core/cmd_search_rop.c (hunk 6 above)
Comment (radare) [inline libr/core/cmd_search_rop.c]: so?

**Pull request merged**